shared/src/main/scala/mlscript/Typer.scala

```diff
@@ -20,60 +20,94 @@ class Typer(var dbg: Boolean, var verbose: Bool, var explainErrors: Bool) extend
   type Raise = Diagnostic => Unit
   type Binding = Str -> TypeScheme
   type Bindings = Map[Str, TypeScheme]
-  
+
   case class TypeDef(
     kind: TypeDefKind,
     nme: TypeName,
-    tparams: List[TypeName],
-    targs: List[TypeVariable],
-    body: Type,
+    tparamsargs: List[(TypeName, TypeVariable)],
+    tvars: Iterable[TypeVariable], // "implicit" type variables. instantiate every time a `TypeRef` is expanded
```

Comment:
I used the return type of `Map.values` (in `typeType2`) to avoid creating unnecessary collections. This would be removed when we fix the soundness issue with type variables.